GPflow/ekernels.py

```diff
@@ -76,17 +76,11 @@ def exKxz(self, Z, Xmu, Xcov):
             tf.expand_dims(eye(tf.shape(Xmu)[1]), 0) + tf.reshape(lengthscales ** -2.0, (1, 1, -1)) * Xsigm
         )  # N
 
-        vec = tf.expand_dims(Z, 0) - tf.expand_dims(Xmum, 1)  # NxMxD
-
-        rsm = tf.tile(tf.expand_dims(scalemat, 1), (1, M, 1, 1))  # Reshaped scalemat
-        smIvec = tf.matrix_solve(rsm, tf.expand_dims(vec, 3))[:, :, :, 0]  # NxMxDx1
-        q = tf.reduce_sum(smIvec * vec, [2])  # NxM
+        vec = tf.expand_dims(tf.transpose(Z), 0) - tf.expand_dims(Xmum, 2)  # NxDxM
+        smIvec = tf.matrix_solve(scalemat, vec)  # NxDxM
+        q = tf.reduce_sum(smIvec * vec, [1])  # NxM
 
-        addvec = tf.matmul(
-            tf.tile(tf.expand_dims(Xsigmc, 1), (1, M, 1, 1)),
-            tf.expand_dims(smIvec, 3),
-            transpose_a=True
-        )[:, :, :, 0] + tf.expand_dims(Xmup, 1)  # NxMxD
+        addvec = tf.matmul(smIvec, Xsigmc, transpose_a=True) + tf.expand_dims(Xmup, 1)  # NxMxD
```

Comment:
Interesting that `tf.einsum` is slightly quicker. For now I would argue we should still use `tf.matmul`, since the difference is small, and the one example doesn't give a super clear image  about what happens when matrices are much larger, or what happens on different devices. `tf.einsum` is very new, and I've recently encountered some bugs in it.

I could be convinced otherwise, but I'd have to understand how both ops work and how they're optimised.